### DISCUSSION TITLE
Track the path in the nested structures

### DISCUSSION THREAD

**Discussion by naquad** [Respect/Validation#1427] [Ideas]
Hello,

I'm evaluating various validation libraries in order to find one that is expressive and able to validate nested structures. Respect is great except that I can't find the path of the failed key. What I mean:

```php
<?php

use Respect\Validation\Exceptions\NestedValidationException;
use Respect\Validation\Validator as v;

$acl_update_request = [
    15 => [
        'groups' => [1, 2, 3, 4, 5],
        'permissions' => [
            'perm1' => true,
            'perm2' => false,
            'perm3' => 'boom!',
        ]
    ]
];

try {
    v::arrayVal()->
        each(v::arrayVal()->each(v::boolVal())->
        each(v::boolVal()))->
        assert($acl_update_request);
} catch(NestedValidationException $exception) {
    var_dump($exception->getMessages());
}
```

I can't get "15.permissions.perm3" in any way. W/o this feature validation loses half of its value. The goal is to get something like:
```php
[
  15 => [
    'permissions' => [
      'perm3' => [
        'must be a boolean value'
      ]
    ]
  ]
]
```

My case is quite a common one: fat forms and complex APIs. If validation can't say exactly what is wrong, the usage and debugging become a nightmare.

Are there any plans for tracking paths?

Thank you.

**Comment by henriquemoody** [comment]
In version 3.0, the array messages will look like this:

```php
[
    'each' => [
        '__root__' => 'These rules must pass for `["groups": [1, 2, 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]]`',
        'each.1' => [
            '__root__' => 'Each item in `["groups": [1, 2, 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]]` must be valid',
            'boolVal.1' => '`[1, 2, 3, 4, 5]` must be a boolean value',
            'boolVal.2' => '`["perm1": true, "perm2": false, "perm3": "boom!"]` must be a boolean value',
        ],
        'each.2' => [
            '__root__' => 'Each item in `["groups": [1, 2, 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]]` must be valid',
            'boolVal.1' => '`[1, 2, 3, 4, 5]` must be a boolean value',
            'boolVal.2' => '`["perm1": true, "perm2": false, "perm3": "boom!"]` must be a boolean value',
        ],
    ],
]
```

What do you think about that, I'm sure it's better, but I would like to know if that's good enough,

**Reply by naquad** [reply]
Unfortunately, it isn't better as I'll have to parse the error messages. The need is to get the exact path of the failed VALUE, not the rule.

**Comment by henriquemoody** [comment]
However, I think what you want is a validation chain that looks like this:

```php
use Respect\Validation\Exceptions\NestedValidationException;
use Respect\Validation\Validator as v;

$acl_update_request = [
    15 => [
        'groups' => [1, 2, 3, 4, 5],
        'permissions' => [
            'perm1' => true,
            'perm2' => false,
            'perm3' => 'boom!',
        ]
    ]
];

try {
    v::arrayVal()
        ->each(
            v::arrayVal()
                ->key('groups', v::arrayVal()->each(v::intVal()))
                ->key('permissions', v::arrayVal()->each(v::boolVal()))
        )
        ->assert($acl_update_request);
} catch(NestedValidationException $exception) {
    var_dump($exception->getMessages());
}
```

That will then result into:
```php
[
    'each' => 'permissions must be a boolean value',
]
```

**Reply by naquad** [reply]
No, it is the undesired behavior as described in the original case.

**Reply by henriquemoody** [reply]
I see, I think it's quite hard to get the paths the way you want, because the validation doesn't aways match the structure. What would you expect if your input was like this?

 ```php
[
    15 => [
        'groups' => [1, 2, 3, 4, 5],
        'permissions' => [
            'perm1' => true,
            'perm2' => false,
            'perm3' => 'boom!',
        ]
    ],
    16 => false,
    17 => [false],
    17 => ['permissions' => false],
]
```

**Reply by naquad** [reply]
Something akin to:

```
[
  15 => [
    'permissions' => [
      'perm3' => [
        'must be a boolean value'
      ]
    ]
  ],
  16 => ['must be an array'],
  17 => ['groups' => ['required'], 'permissions' => ['required'], 0 => ['invalid key']],
  18 => ['groups' => ['required'], 'permissions' => ['must be an array']],
]
```

If something is invalid, then obviously, its children can't be validated. There is no need to completely match the structure; you can "grow" a new structure such that its leafs will be errors.

**Reply by henriquemoody** [reply]
I don't think we can do this natively how the library works, because the validation doesn't know the path of what's validating.

For example, the "each" rule doesn't care about they keys of the array it validates, hence it's impossible to get that key right. I think you could only solve that by iterating over the input, and validating each value manually.

**Reply by naquad** [reply]
Yes, this is a 3 y.o. discovery :) While the library doesn't know what it's validating, the user wants to know where the error is. Having an error "each: value is invalid" for an invalid mile-long object built from a complex form is no good, so while the function of the validation has been satisfied the user entering the data is definitely not.

**Reply by henriquemoody** [reply]
Did you find anything that met your requirements?

**Reply by naquad** [reply]
Hacked on top of Laravel's validator. Lots and lots of hacks with some trade-offs when it came to the data structures.

**Reply by henriquemoody** [reply]
Is that something you could share with me? Perhaps I could use that as inspiration.

Also, what's your use case? I assume it's an API request validation. I'm asking because I can see how useful that would be on a form, and that this library can provide that. When it comes to those complex validations like yours, I wonder how useful it would be to even have the validation the way you want it to be.

**Reply by naquad** [reply]
That won't help much. The reason is that's a ton of imperative code in the Validator's post-processing. Here's the [key FormRequest class](https://gist.github.com/naquad/215754579b52236dcb64a5c713ea0002). Mind how the error's are added manually (`level1.level2.level3 => error`), by default Valitor is not able to construct such errors. Obviously, this is an ad-hoc code and I would really love to replace it with something better.

**Reply by henriquemoody** [reply]
Oh yeah, that's exactly what I meant with _I think you could only solve that by iterating over the input_. It's quite hard to make this generic, possible, but quite hard. I'm gonna play with that later, maybe I come up with something. I will keep you posted.

**Reply by henriquemoody** [reply]
I've played around with it a little bit, so far I've gotten something like this:

PHP code.
```php
$input = [
        16 => [
            'groups' => [1, 'A', 3, 4, 5],
            'permissions' => [
                'perm1' => true,
                'perm2' => false,
                'perm3' => 'boom!',
            ]
        ],
        18 => false,
        24 => ['permissions' => false],
    ];
    v::each(
            v::key('groups', v::each(v::intVal()))
                ->key('permissions', v::each(v::boolVal()))
        )
        ->assert($input);
```

Array messages
```
[
    '__root__' => 'Each item in `[16: ["groups": [1, "A", 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]], 18: false, ... ]` must be valid',
    16 => [
        '__root__' => 'All of the required rules must pass for `["groups": [1, "A", 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]]`',
        'groups' => [
            1 => 'groups must be an integer value',
        ],
        'permissions' => [
            'perm3' => 'permissions must be a boolean value',
        ],
    ],
    18 => [
        '__root__' => 'All of the required rules must pass for `false`',
        'groups' => 'groups must be present',
        'permissions' => 'permissions must be present',
    ],
    24 => [
        '__root__' => 'All of the required rules must pass for `["permissions": false]`',
        'groups' => 'groups must be present',
        'permissions' => 'permissions must be iterable',
    ],
]
```

It caused quite some side effects, but I'm trying to deal with them.

What do you think about that, so far?

**Reply by naquad** [reply]
Looks very promising! Although what happens if the user passes object that has __root__ key? Wouldn't there be a naming conflict? I mean the validator must report that __root__ key is invalid.

**Reply by henriquemoody** [reply]
Yeah, it there could be a conflict if the key is called `__root__`.  I think it's a bit unlikely, but it's possible indeed. I'm not sure how to solve that problem, and I don't want to ignore this message in some cases.

**Reply by henriquemoody** [reply]
Ref https://github.com/Respect/Validation/pull/1498

**Reply by naquad** [reply]
Maybe change the structure so that children and parent messages are separated?

```php
[
   'messages' => ['Each item in `[16: ["groups": [1, "A", 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]], 18: false, ... ]` must be valid'],
   'children' => [
        16 => [
            'messages' => ['All of the required rules must pass for `["groups": [1, "A", 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]]`'],
            'children' => [
              'groups' => [
                  'messages' => ['All of the ...'],
                  'children' => [
                      'messages' => ['groups must be an integer value'],
                  ]
              ],
              'permissions' => [
                  'messages' => ['All of the ...'],
                  'children' => [
                      'perm3' => [
                          'messages' => 'permissions must be a boolean value'
                      ],
                  ]
              ],
            ]
        ],
        // ...
    ]
]

```

While it is more clumsy, no name conflicts may occur. However, naming needs an improvement.

**Reply by henriquemoody** [reply]
That's a way to do it, but it will make the arrays more verbose. Would that be better?

**Reply by henriquemoody** [reply]
Also, the end user adding a `__root__` won't be a problem, the problem is if the developer using the library _wants_ someone to add a key named `__root__`.

**Reply by naquad** [reply]
Verbosity doesn't matter much as long as one can reconstruct the data path so that errors can be located in the structure. More nesting will make things somewhat clumsier but it is better to be clumsy and bullet-proof than fluent and incorrect.

**Reply by henriquemoody** [reply]
I see your point. I wouldn't call it incorrect, though, it can be simply a limitation of the library. Certainly that would be more future-proof, but it will also make harder to iterate over messages. Something to think about

**Reply by henriquemoody** [reply]
I've made some substantial changes in the library, I would appreciate if you give your feedback about this file https://github.com/Respect/Validation/pull/1498/files#diff-5dff648d13ed812358dc53c84ea26f4f307c63ba92019953614d4526f6795235

You don't need to review the whole code, although that would be appreciated!

**Reply by naquad** [reply]
Whoa! From what I see, there are a huge number of changes. Thank you for the effort you made to handle this case. The error structure looks good. I will try to rewrite my code using the Validation on Monday and see where it gets me.

**Reply by henriquemoody** [reply]
I came up with the structure that I'm happy about, it was spired by one of your comments. I think that you're right, it is indeed better to have a lengthier array than use conventions or hide information, so here's where I'm at at the moment:

```php
[
    'messages' => ['each' => 'Each item in `[16: ["groups": [1, "A", 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]], 18: false, ... ]` must be valid'],
    'children' => [
        16 => [
            'messages' => ['allOf' => '`["groups": [1, "A", 3, 4, 5], "permissions": ["perm1": true, "perm2": false, "perm3": "boom!"]]` must pass the rules'],
            'children' => [
                'groups' => [
                    'messages' => ['each' => 'Each item in `[1, "A", 3, 4, 5]` must be valid'],
                    'children' => [
                        1 => [
                            'messages' => ['intVal' => '"A" must be an integer value'],
                        ],
                    ],
                ],
                'permissions' => [
                    'messages' => ['each' => 'Each item in `["perm1": true, "perm2": false, "perm3": "boom!"]` must be valid'],
                    'children' => [
                        'perm3' => [
                            'messages' => ['boolVal' => '"boom!" must be a boolean value'],
                        ],
                    ],
                ],
            ],
        ],
        18 => [
            'messages' => ['allOf' => '`false` must pass all the rules'],
            'details' => ['arrayVal' => '`false` must be an array value'],
            'children' => [
                'groups' => [
                    'messages' => ['keyExists' => '`.groups` must be present'],
                ],
                'permissions' => [
                    'messages' => ['keyExists' => '`.permissions` must be present'],
                ],
            ],
        ],
        24 => [
            'messages' => ['allOf' => '`["permissions": false]` must pass the rules'],
            'children' => [
                'groups' => [
                    'messages' => ['keyExists' => '`.groups` must be present'],
                ],
                'permissions' => [
                    'messages' => ['each' => '`false` must be iterable'],
                ],
            ],
        ],
    ],
]
```

A few things to note here:

1. There's also a key `details` because some rules have children but they're respective to the parent validation, not a child. I thought I would add an extra key there to identify those. I think that one would always use `details` instead of `messages` when it's available because `messages` would usually be only headers of nesting validations.
2. I'm not replacing the input of each rule with the key, because I think that, since the structure is already in the array, it's best to keep the original message instead, so you can actually see the value that failed.

What are your thoughts?

**Reply by naquad** [reply]
It looks great! The only thing is the details: will its presence *always* imply a failed parent validation that cascades further, or should there be some logic handling special cases? In other words, the question is where to stop the error deducer: on the details if present and the whole key has an issue, or should it traverse the children errors despite the details presence?

**Reply by henriquemoody** [reply]
It always implies that the current level failed validation. But whether you still go down the children will depend on the approach you take.

Personally, for your use case, I wouldn't add the `arrayVal` to the chain because `key` would check that anyway, and for the user it's more useful to know already which keys they need to provide rather than knowing they need to provide an array.

**Reply by naquad** [reply]
Ah, that makes things clear. Thank you again :) Is this version available in some commits? I want to try it out and see where it gets me.

**Reply by henriquemoody** [reply]
The version is only available on my merge request (or my branch https://github.com/henriquemoody/Validation/tree/core/path). I might make some changes, so maybe you could fork my fork, and add the a VCS repository to your composer.json file?
